### PR TITLE
boards: arm: nrf5340_dk: Enable DC/DC by default & Fix NRF_CACHE and NVMC macro defines

### LINKED ISSUE #20695: nRF5340: misc fixes for nRF53 porting
**Describe the bug**
Several bugs need to be fixed for the nRF53 SoC porting in Zephyr

- peripheral macro definitions for CACHE and NVMC need to be corrected. Instead of directly referencing CACHE_S and NVMC_S macros (i.e. for secure domain), the NRF_CACHE and NRF_NVMC macros may be used directly. This has been addressed in nRFX, where the correct base peripheral addresses are generated based on the security domain of the image we are building.

- DCDC support, present in the nRF51 and nRF52 series of SoC/Boards, is a default nRF board feature, however, is missing for nRF53 and needs to be added so the nRF53-based Boards have the same set of default features.

### PR DESCRIPTION
**Pull request by cvinayak**
Since the nRF5340 DK contains the required DC/DC circuitry,
enable it by default while keeping it configurable. This
reduces power consumption in application core , network
core and high voltage use.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Fixes #20695

### CODE REVIEW

**Review by ioannisg — CHANGES_REQUESTED** [review]
Thanks for the patch @cvinayak .
Note however it won't work for App mcu executing on non secure mode
We need to have a way to access regulators from NS domain or just don't do it at all.

**Comment by cvinayak** [comment]
I just followed the existing code for NRF_CACHE_S, how different should it be that soc.c can handle NS mode? Feel free to suggest changes.

**Comment by ioannisg** [comment]
> I just followed the existing code for NRF_CACHE_S, how different should it be that soc.c can handle NS mode? Feel free to suggest changes.

Oh, right, I understand your confusion. It should not be different from CACHE, this seems to be a minor issue IMO.

So, `SOC_NRF5340_CPUAPP `is selected for both Secure and Non-Secure builds for the Application MCU. Therefore, any access to _S needs to be guarded not only by `CONFIG_SOC_NRF5340_CPUAPP`, but also with `!CONFIG_TRUSTED_EXECUTION_NONSECURE`.

The reason it works for nRF53 is that the whole thing is guarded by CONFIG_NRF_ENABLE_CACHE, and _that_ is defined as.
```
config NRF_ENABLE_CACHE
	bool "Enable cache"
	depends on (SOC_NRF5340_CPUAPP && !TRUSTED_EXECUTION_NONSECURE) \
			|| SOC_NRF5340_CPUNET
	default y
```
...i.e. the guarding with !TRUSTED_EXECUTION_NONSECURE occurs already at the Kconfig definition.

The reason I did it this way is that I did not want the `NRF_ENABLE_CACHE `option to be enabled at all for Non-Secure builds.

To summarize. what I suggest is to guard your REGULATORS_S aditionally with !defined(CONFIG_TRUSTED_EXECUTION_NONSECURE) and if you want, you can add the same guard around CACHE_S.

**Review by ioannisg — CHANGES_REQUESTED** [review]
Actually, please, disregard my comment, @cvinayak seems we have a way to pull the right NRF_xxx macro via nrfx, now (enlighted by @anangl ).
I think you can simply use NRF_REGULATORS, then. Not sure if this is present on Network MCU.

File: boards/arm/nrf5340_dk_nrf5340/Kconfig [hunk 1]
```diff
@@ -5,6 +5,21 @@
 
 if BOARD_NRF5340_DK_NRF5340_CPUAPP || BOARD_NRF5340_DK_NRF5340_CPUAPPNS
 
+config BOARD_ENABLE_DCDC_APP
+	bool "Enable Application Core DCDC converter"
```
Comment (ioannisg) [inline boards/arm/nrf5340_dk_nrf5340/Kconfig]: Please, use "MCU" instead of "Core" ; this is the name we use consistently in doc/index.rst.

File: soc/arm/nordic_nrf/nrf53/Kconfig.soc [hunk 2]
```diff
@@ -89,6 +89,21 @@ config SOC_NRF5340_CPUNET_QKAA
 
 endchoice
 
+config SOC_DCDC_NRF53X_APP
+	bool
+	help
+	  Enable nRF53 series System on Chip Application Core DC/DC converter.
```
Comment (ioannisg) [inline soc/arm/nordic_nrf/nrf53/Kconfig.soc]: PLease, use "MCU" instead of Core, here, and below

File: boards/arm/nrf5340_dk_nrf5340/Kconfig (hunk 1 above)
Comment (ioannisg) [reply boards/arm/nrf5340_dk_nrf5340/Kconfig]: and in similar places below

File: soc/arm/nordic_nrf/nrf53/Kconfig.soc [hunk 3]
```diff
@@ -89,6 +89,21 @@ config SOC_NRF5340_CPUNET_QKAA
 
 endchoice
 
+config SOC_DCDC_NRF53X_APP
```
Comment (ioannisg) [inline soc/arm/nordic_nrf/nrf53/Kconfig.soc]: Can you add dependencies in these options, i.e. depend on SOC_NRF5340_CPUAPP

File: soc/arm/nordic_nrf/nrf53/soc.c [hunk 4]
```diff
@@ -59,6 +59,18 @@ static int nordicsemi_nrf53_init(struct device *arg)
 	*((u32_t *)0x500046D0) = 0x1;
 #endif
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP)
+#if defined(CONFIG_SOC_DCDC_NRF53X_APP)
+	NRF_REGULATORS_S->VREGMAIN.DCDCEN = 1;
```
Comment (ioannisg) [inline soc/arm/nordic_nrf/nrf53/soc.c]: Just try to use NRF_REGULATORS here, i.e. without _S

Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf53/soc.c]: and below

**Review by ioannisg — COMMENTED** [review]
@anangl could you take a look as well?

File: soc/arm/nordic_nrf/nrf53/soc.c [hunk 5]
```diff
@@ -59,6 +59,18 @@ static int nordicsemi_nrf53_init(struct device *arg)
 	*((u32_t *)0x500046D0) = 0x1;
 #endif
 
+#if defined(CONFIG_SOC_NRF5340_CPUAPP)
```
Comment (anangl) [inline soc/arm/nordic_nrf/nrf53/soc.c]: This does not seem to be needed.

File: soc/arm/nordic_nrf/nrf53/Kconfig.soc (hunk 3 above)
Comment (cvinayak) [reply soc/arm/nordic_nrf/nrf53/Kconfig.soc]: done

**Review by anangl — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]
